File: app/src/main/java/org/thoughtcrime/securesms/conversation/ConversationReactionDelegate.java
package org.thoughtcrime.securesms.conversation;

import android.app.Activity;
import android.graphics.PointF;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.NonNull;

import org.thoughtcrime.securesms.database.model.MessageRecord;
import org.thoughtcrime.securesms.recipients.Recipient;
import org.thoughtcrime.securesms.util.views.Stub;

/**
 * Delegate class that mimics the ConversationReactionOverlay public API
 * <p>
 * This allows us to properly stub out the ConversationReactionOverlay View class while still
 * respecting listeners and other positional information that can be set BEFORE we want to actually
 * resolve the view.
 */
public final class ConversationReactionDelegate {

  private final Stub<ConversationReactionOverlay> overlayStub;
  private final PointF                            lastSeenDownPoint = new PointF();

  private ConversationReactionOverlay.OnReactionSelectedListener onReactionSelectedListener;
  private ConversationReactionOverlay.OnActionSelectedListener   onActionSelectedListener;
  private ConversationReactionOverlay.OnHideListener             onHideListener;

  public ConversationReactionDelegate(@NonNull Stub<ConversationReactionOverlay> overlayStub) {
    this.overlayStub = overlayStub;
  }

  public boolean isShowing() {
    return overlayStub.resolved() && overlayStub.get().isShowing();
  }

  public void show(@NonNull Activity activity,
                   @NonNull Recipient conversationRecipient,
                   @NonNull ConversationMessage conversationMessage,
                   boolean isNonAdminInAnnouncementGroup,
                   @NonNull SelectedConversationModel selectedConversationModel,
                   boolean canEditGroupInfo)
  {
    resolveOverlay().show(activity, conversationRecipient, conversationMessage, lastSeenDownPoint, isNonAdminInAnnouncementGroup, selectedConversationModel, canEditGroupInfo);
  }

  public void hide() {
    overlayStub.get().hide();
  }

  public void setOnReactionSelectedListener(@NonNull ConversationReactionOverlay.OnReactionSelectedListener onReactionSelectedListener) {
    this.onReactionSelectedListener = onReactionSelectedListener;

    if (overlayStub.resolved()) {
      overlayStub.get().setOnReactionSelectedListener(onReactionSelectedListener);
    }
  }

  public void setOnActionSelectedListener(@NonNull ConversationReactionOverlay.OnActionSelectedListener onActionSelectedListener) {
    this.onActionSelectedListener = onActionSelectedListener;

    if (overlayStub.resolved()) {
      overlayStub.get().setOnActionSelectedListener(onActionSelectedListener);
    }
  }

  public void setOnHideListener(@NonNull ConversationReactionOverlay.OnHideListener onHideListener) {
    this.onHideListener = onHideListener;

    if (overlayStub.resolved()) {
      overlayStub.get().setOnHideListener(onHideListener);
    }
  }

  public @NonNull MessageRecord getMessageRecord() {
    if (!overlayStub.resolved()) {
      throw new IllegalStateException("Cannot call getMessageRecord right now.");
    }

    return overlayStub.get().getMessageRecord();
  }

  public boolean applyTouchEvent(@NonNull MotionEvent motionEvent) {
    if (!overlayStub.resolved() || !overlayStub.get().isShowing()) {
      if (motionEvent.getAction() == MotionEvent.ACTION_DOWN) {
        lastSeenDownPoint.set(motionEvent.getX(), motionEvent.getY());
      }
      return false;
    } else {
      return overlayStub.get().applyTouchEvent(motionEvent);
    }
  }

  private @NonNull ConversationReactionOverlay resolveOverlay() {
    boolean                     wasAlreadyResolved = overlayStub.resolved();
    ConversationReactionOverlay overlay            = overlayStub.get();

    if (!wasAlreadyResolved && (overlay.getWidth() == 0 || overlay.getHeight() == 0)) {
      // force immediate measurement and layout after ViewStub inflation to ensure proper dimensions before first use
      // without doing this, the overlay child views will be positioned off screen in RTL layout direction because of negative values.

      View parent = (View) overlay.getParent();
      if (parent != null) {
        int widthSpec  = View.MeasureSpec.makeMeasureSpec(parent.getWidth(), View.MeasureSpec.EXACTLY);
        int heightSpec = View.MeasureSpec.makeMeasureSpec(parent.getHeight(), View.MeasureSpec.EXACTLY);
        overlay.measure(widthSpec, heightSpec);
        overlay.layout(0, 0, overlay.getMeasuredWidth(), overlay.getMeasuredHeight());
      }
    }

    overlay.requestFitSystemWindows();

    overlay.setOnHideListener(onHideListener);
    overlay.setOnActionSelectedListener(onActionSelectedListener);
    overlay.setOnReactionSelectedListener(onReactionSelectedListener);

    return overlay;
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/conversation/v2/ConversationReactionDelegate.kt
package org.thoughtcrime.securesms.conversation.v2

import android.app.Activity
import android.graphics.PointF
import android.view.MotionEvent
import org.session.libsession.utilities.Stub
import org.thoughtcrime.securesms.conversation.v2.ConversationReactionOverlay.OnActionSelectedListener
import org.thoughtcrime.securesms.conversation.v2.ConversationReactionOverlay.OnHideListener
import org.thoughtcrime.securesms.conversation.v2.ConversationReactionOverlay.OnReactionSelectedListener
import org.thoughtcrime.securesms.database.model.MessageRecord

/**
 * Delegate class that mimics the ConversationReactionOverlay public API
 *
 * This allows us to properly stub out the ConversationReactionOverlay View class while still
 * respecting listeners and other positional information that can be set BEFORE we want to actually
 * resolve the view.
 */
internal class ConversationReactionDelegate(private val overlayStub: Stub<ConversationReactionOverlay>) {
    private val lastSeenDownPoint = PointF()
    private var onReactionSelectedListener: OnReactionSelectedListener? = null
    private var onActionSelectedListener: OnActionSelectedListener? = null
    private var onHideListener: OnHideListener? = null
    val isShowing: Boolean
        get() = overlayStub.resolved() && overlayStub.get().isShowing

    fun show(
        activity: Activity,
        messageRecord: MessageRecord,
        selectedConversationModel: SelectedConversationModel,
        blindedPublicKey: String?
    ) {
        resolveOverlay().show(
            activity,
            messageRecord,
            lastSeenDownPoint,
            selectedConversationModel,
            blindedPublicKey
        )
    }

    fun hide() {
        overlayStub.get().hide()
    }

    fun hideForReactWithAny() {
        overlayStub.get().hideForReactWithAny()
    }

    fun setOnReactionSelectedListener(onReactionSelectedListener: OnReactionSelectedListener) {
        this.onReactionSelectedListener = onReactionSelectedListener
        if (overlayStub.resolved()) {
            overlayStub.get().setOnReactionSelectedListener(onReactionSelectedListener)
        }
    }

    fun setOnActionSelectedListener(onActionSelectedListener: OnActionSelectedListener) {
        this.onActionSelectedListener = onActionSelectedListener
        if (overlayStub.resolved()) {
            overlayStub.get().setOnActionSelectedListener(onActionSelectedListener)
        }
    }

    fun setOnHideListener(onHideListener: OnHideListener) {
        this.onHideListener = onHideListener
        if (overlayStub.resolved()) {
            overlayStub.get().setOnHideListener(onHideListener)
        }
    }

    val messageRecord: MessageRecord
        get() {
            check(overlayStub.resolved()) { "Cannot call getMessageRecord right now." }
            return overlayStub.get().messageRecord
        }

    fun applyTouchEvent(motionEvent: MotionEvent): Boolean {
        return if (!overlayStub.resolved() || !overlayStub.get().isShowing) {
            if (motionEvent.action == MotionEvent.ACTION_DOWN) {
                lastSeenDownPoint[motionEvent.x] = motionEvent.y
            }
            false
        } else {
            overlayStub.get().applyTouchEvent(motionEvent)
        }
    }

    private fun resolveOverlay(): ConversationReactionOverlay {
        val overlay = overlayStub.get()
        overlay.requestFitSystemWindows()
        overlay.setOnHideListener(onHideListener)
        overlay.setOnActionSelectedListener(onActionSelectedListener)
        overlay.setOnReactionSelectedListener(onReactionSelectedListener)
        return overlay
    }
}